Index: linux/src/drivers/block/ide.h
===================================================================
RCS file: /cvsroot/hurd/gnumach/linux/src/drivers/block/Attic/ide.h,v
retrieving revision 1.1
diff -u -p -r1.1 ide.h
--- linux/src/drivers/block/ide.h	26 Apr 1999 05:51:39 -0000	1.1
+++ linux/src/drivers/block/ide.h	29 Dec 2007 18:09:54 -0000
@@ -23,7 +23,7 @@
  * REALLY_SLOW_IO can be defined in ide.c and ide-cd.c, if necessary
  */
 #undef REALLY_FAST_IO			/* define if ide ports are perfect */
-#define INITIAL_MULT_COUNT	0	/* off=0; on=2,4,8,16,32, etc.. */
+#define INITIAL_MULT_COUNT	16	/* off=0; on=2,4,8,16,32, etc.. */
 
 #ifndef SUPPORT_SLOW_DATA_PORTS		/* 1 to support slow data ports */
 #define SUPPORT_SLOW_DATA_PORTS	1	/* 0 to reduce kernel size */
